x/gamm/keeper/msg_server.go

```diff
@@ -309,3 +315,50 @@ func (server msgServer) ExitSwapShareAmountIn(goCtx context.Context, msg *types.
 
 	return &types.MsgExitSwapShareAmountInResponse{TokenOutAmount: tokenOutAmount}, nil
 }
+
+func (server msgServer) MigrateShares(goCtx context.Context, msg *balancer.MsgMigrateSharesToFullRangeConcentratedPosition) (amount0, amount1 sdk.Int, liquidity sdk.Dec, err error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	sender, err := sdk.AccAddressFromBech32(msg.Sender)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, err
+	}
+
+	// get the poolId by parsing the gamm share denom
+	poolIdLeaving, err := GetPoolIdFromSharesDenom(msg.SharesToMigrate.Denom)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, err
+	}
+
+	exitCoins, err := server.keeper.ExitPool(ctx, sender, poolIdLeaving, msg.SharesToMigrate.Amount, sdk.NewCoins())
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, err
+	}
+
+	pool, err := server.keeper.clKeeper.GetPoolById(ctx, msg.PoolIdEntering)
```

Comment:
It might make sense to first try to get the CL pool before exiting the gamm pool in case the CL pool ends up being non-existent

Reply:
Done!

https://github.com/osmosis-labs/osmosis/pull/4051/commits/f0fccf345f46f4392c778accd153dacef96620e9